operator/pkg/ciliumenvoyconfig/manager.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return nil, err
+	}
+	manager.envoyConfigManager = envoyConfigManager
+
+	go manager.run(ctx)
+	return manager, nil
+}
+
+func (m *manager) OnAddService(service *slim_corev1.Service) error {
+	var (
+		svcName   = service.Name
+		scopedLog = log.WithField(logfields.ServiceName, svcName)
+	)
+	key, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		return err
+	}
+
+	if !IsLBProtocolAnnotationEnabled(service) && !hasAnyPort(service, m.ports) {
+		return nil
+	}
+
+	scopedLog.Debug("adding event to queue")
+	m.queue.Add(svcEvent(key))
+	return nil
+}
+
+func hasAnyPort(svc *slim_corev1.Service, ports []string) bool {
+	for _, p := range ports {
+		for _, port := range svc.Spec.Ports {
+			if p == getServiceFrontendPort(port) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getServiceFrontendPort(port slim_corev1.ServicePort) string {
+	if port.Port != 0 {
+		return strconv.Itoa(int(port.Port))
+	}
+	if port.NodePort != 0 {
+		return strconv.Itoa(int(port.NodePort))
+	}
+	return port.Name
+}
+
+func (m *manager) OnUpdateService(oldObj, newObj *slim_corev1.Service) error {
+	var (
+		svcName   = newObj.Name
+		scopedLog = log.WithField(logfields.ServiceName, svcName)
+	)
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		return err
+	}
+
+	if !IsLBProtocolAnnotationEnabled(newObj) && !IsLBProtocolAnnotationEnabled(oldObj) &&
+		!hasAnyPort(newObj, m.ports) && !hasAnyPort(oldObj, m.ports) {
+		return nil
+	}
+	scopedLog.Debug("adding event to queue")
+	m.queue.Add(svcEvent(key))
+	return nil
+}
+
+func (m *manager) OnDeleteService(_ *slim_corev1.Service) error {
+	// Doing nothing here as clean up should be done via k8s OwnerReferences
+	return nil
+}
+
+// run kicks off the controlled loop
+func (m *manager) run(ctx context.Context) {
+	defer m.queue.ShutDown()
+	for {
+		ev, quit := m.queue.Get()
+		if quit {
+			return
+		}
+		err := m.processEvent(ctx, ev)
+		if err != nil {
+			if m.queue.NumRequeues(ev) < m.maxRetries {
+				log.WithError(err).Warning("Error while processing event, retrying")
+				m.queue.AddRateLimited(ev)
+			} else {
+				log.WithError(err).Warning("Error while processing event, no more retries")
+				m.queue.Forget(ev)
+			}
+			continue
+		}
+		m.queue.Done(ev)
+	}
+}
+
+func (m *manager) processEvent(ctx context.Context, event interface{}) error {
+	switch k := event.(type) {
+	case svcEvent:
+		n := string(k) // service namespace/name
+
+		objFromCache, exists, err := m.serviceStore.GetByKey(n)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("service %s does not exist", n)
+		}
+
+		service, ok := objFromCache.(*slim_corev1.Service)
+		if !ok {
+			return fmt.Errorf("got invalid object from cache: %T", objFromCache)
+		}
+		if IsLBProtocolAnnotationEnabled(service) || hasAnyPort(service, m.ports) {
+			return m.createEnvoyConfig(ctx, service)
+		}
+		return m.deleteEnvoyConfig(ctx, service)
```

Comment:
This is a bit confusing IMO, I think we should probably have a condition where we call delete, instead of it being the fallback. Is this intending to delete the config when the service is deleted? Perhaps we should check the deletionTimestamp before deleting.

Reply:
this is for the case user add or remove or update the annotation.